### PR TITLE
Fix pre-activation implementation; add regularization 

### PR DESCRIPTION
**Pull request by JefferyRPrice**
- Addresses issues #25 and #26 by adding L2 regularization (weight decay) and using pre-activations according to paper
- Changes optimizer from RMSprop to Adam (see https://github.com/fchollet/keras/issues/4897)
https://github.com/raghakot/keras-resnet/issues/26
- Adds learning rate scheduler to cifar10.py
- Achieves about 87.8% in cifar10.py experiment
- Note that ResNet18 as implemented doesn't really seem appropriate for CIFAR-10 as the last two residual stages end up as all 1x1 convolutions from downsampling (stride). This is worse for deeper versions. Using a smaller, modified [ResNet-like architecture](https://gist.github.com/JefferyRPrice/c1ecc3d67068c8d9b3120475baba1d7e), I was able to achieve 92% on CIFAR-10.

### CODE REVIEW

File: resnet.py [hunk 1]
```diff
@@ -12,16 +12,21 @@
...
+
 
 # Helper to build a conv -> BN -> relu block
 def _conv_bn_relu(nb_filter, nb_row, nb_col, subsample=(1, 1)):
     def f(input):
         conv = Convolution2D(nb_filter=nb_filter, nb_row=nb_row, nb_col=nb_col, subsample=subsample,
-                             init="he_normal", border_mode="same")(input)
-        norm = BatchNormalization(mode=0, axis=CHANNEL_AXIS)(conv)
-        return Activation("relu")(norm)
+                             init="he_normal", border_mode="same", W_regularizer=l2(0.0001))(input)
```
Comment (raghakot) [inline resnet.py]: instead of hardcoding these, we should allow conv params to be passed as **conv_params

File: cifar10.py [hunk 2]
```diff
@@ -33,7 +49,7 @@
 
 # Let's train the model using RMSprop
```
Comment (raghakot) [inline cifar10.py]: Its no longer rmsprop. update or remove comment

File: cifar10.py [hunk 3]
```diff
@@ -6,9 +6,25 @@
 from keras.datasets import cifar10
 from keras.preprocessing.image import ImageDataGenerator
 from keras.utils import np_utils
+from keras.callbacks import LearningRateScheduler
 
 import resnet
 
+def _lr_schedule(epoch):
```
Comment (raghakot) [inline cifar10.py]: Why not just use ReduceLROnPlateau>=?

File: resnet.py (hunk 1 above)
Comment (JefferyRPrice) [inline resnet.py]: Done. FWIW, I don't really think this makes the builder any more flexible since it's very specifically targeted to build just the architectures from the papers, rather than to allow those architectures as special cases of a more generic builder. Not that that's really a problem, just why I didn't try to be more flexible the first time around.

**Comment by JefferyRPrice** [comment]
As implemented now, ResNet18, with `ReduceLROnPlateau` and `EarlyStopping`, achieves about 86.7% val_acc via `cifar10.py`. Results CSV is attached as is the architecture via `keras.utils.visualize_util.plot`. Note in the architecture that the later stages of ResNet18 aren't really appropriate for CIFAR-10 sized imagery.
resnet18_cifar10_results
resnet18_architecture
[resnet18_cifar10.csv.zip](https://github.com/raghakot/keras-resnet/files/707075/resnet18_cifar10.csv.zip)

**Comment by raghakot** [comment]
Cool. Thanks for the PR.
